### PR TITLE
Added a router listener

### PR DESCRIPTION
**Pull request by Nyholm**
This caches all the result for the router to be used for later. 

TODO: We need to improve the cache key.

### CODE REVIEW

**Comment by Nyholm** [comment]
This implementation caches the response from the router. It requires ~~2~~ 1 cache lookups per HTTP request. When we get a cache miss we spend some extra time on the route lookup because we do not cache the route collection anymore. 

Any comments?

**Comment by Nyholm** [comment]
Do we need the query params in the cache key?

File: src/Routing/RouterListener.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+     * After the routing, make sure to store the result in cache.
+     *
+     * @param GetResponseEvent $event
+     */
+    public function onAfterRouting(GetResponseEvent $event)
+    {
+        $request = $event->getRequest();
+
+        if (!$request->attributes->has('_controller')) {
+            // routing has not taken place
```
Comment (cryptiklemur) [inline src/Routing/RouterListener.php:70]: if it has controller, i think routing HAS taken place, no?

File: src/Routing/RouterListener.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function createCacheKey(Request $request)
+    {
+        $key = sprintf('route:%s:%s:%s',$request->getMethod(),$request->getHost(),$request->getPathInfo());
+
+        // This might be optional
+        //$key.=':'.$this->implodeRecursive('|', $request->query->all());
```
Comment (cryptiklemur) [inline src/Routing/RouterListener.php:98]: I don't think we want this, but we may want route parameters

File: src/Routing/RouterListener.php (hunk 1 above)
Comment (Nyholm) [inline src/Routing/RouterListener.php:70]: You missed the "!". 

If NOT the $request has _controller, no routing has taken place

File: src/Routing/RouterListener.php (hunk 2 above)
Comment (Nyholm) [inline src/Routing/RouterListener.php:98]: Do we know about route parameters before the route?

File: src/Routing/RouterListener.php (hunk 1 above)
Comment (cryptiklemur) [inline src/Routing/RouterListener.php:70]: Derp, i thought this was in onBeforeRouting for some reason.

File: src/Routing/RouterListener.php (hunk 2 above)
Comment (cryptiklemur) [inline src/Routing/RouterListener.php:98]: I'm not sure

Comment (cryptiklemur) [inline src/Routing/RouterListener.php:98]: I dont think so...

Comment (Nyholm) [inline src/Routing/RouterListener.php:98]: I dont think we know that either. The route parameters are set by the router.

File: src/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -142,10 +142,7 @@ private function addRouterSection()
                     ->defaultFalse()
                 ->end()
                 ->integerNode('ttl')
-                    ->defaultValue(86400)
-                ->end()
-                ->booleanNode('auto-register')
-                    ->defaultTrue()
+                    ->defaultValue(604800)
```
Comment (Nyholm) [inline src/DependencyInjection/Configuration.php:145]: Is a week a good default?

File: src/DependencyInjection/CacheExtension.php [hunk 4]
```diff
@@ -45,6 +45,14 @@ public function load(array $configs, ContainerBuilder $container)
...
         }
 
+        if ($config['router']['enabled']) {
+            $container->register('cache.router_listener', RouterListener::class)
+                ->addArgument(new Reference($config['router']['service_id']))
+                ->addArgument($config['router']['ttl'])
+                ->addTag('kernel.event_listener', ['event'=>'kernel.request', 'method'=>'onBeforeRouting', 'priority'=>33])
+                ->addTag('kernel.event_listener', ['event'=>'kernel.request', 'method'=>'onAfterRouting', 'priority'=>31]);
+        }
+
```
Comment (Nyholm) [inline src/DependencyInjection/CacheExtension.php:52]: Maybe consider using a services.xml

File: src/DependencyInjection/Configuration.php (hunk 3 above)
Comment (cryptiklemur) [inline src/DependencyInjection/Configuration.php:145]: Yeah

File: src/DependencyInjection/CacheExtension.php (hunk 4 above)
Comment (cryptiklemur) [inline src/DependencyInjection/CacheExtension.php:52]: Yeah

**Comment by Nyholm** [comment]
Done!

**Pull request merged**